### PR TITLE
Arms memory management

### LINKED ISSUE #75251: Flaky test create_and_destroy: /obj/item/bodypart/arm/left/golem hard deleted 1 times out of a total del count of 7
Flaky tests were detected in [this test run](https://github.com/tgstation/tgstation/actions/runs/4910596115/attempts/1). This means that there was a failure that was cleared when the tests were simply restarted.
Failures:
```
create_and_destroy: /obj/item/bodypart/arm/left/golem hard deleted 1 times out of a total del count of 7 at code/modules/unit_tests/create_and_destroy.dm:187
```

### LINKED ISSUE #75376: Flaky test create_and_destroy: /obj/item/bodypart/arm/right/golem hard deleted 1 times out of a total del count of 7
Flaky tests were detected in [this test run](https://github.com/tgstation/tgstation/actions/runs/4952151040/attempts/1). This means that there was a failure that was cleared when the tests were simply restarted.
Failures:
```
create_and_destroy: /obj/item/bodypart/arm/right/golem hard deleted 1 times out of a total del count of 7 at code/modules/unit_tests/create_and_destroy.dm:185
```

### LINKED ISSUE #75377: Flaky test create_and_destroy: /obj/item/bodypart/arm/left/golem hard deleted 1 times out of a total del count of 8
Flaky tests were detected in [this test run](https://github.com/tgstation/tgstation/actions/runs/4953020368/attempts/1). This means that there was a failure that was cleared when the tests were simply restarted.
Failures:
```
create_and_destroy: /obj/item/bodypart/arm/left/golem hard deleted 1 times out of a total del count of 8 at code/modules/unit_tests/create_and_destroy.dm:185
```

### LINKED ISSUE #75394: Multiple errors in flaky test create_and_destroy
Flaky tests were detected in [this test run](https://github.com/tgstation/tgstation/actions/runs/4960980280/attempts/1). This means that there was a failure that was cleared when the tests were simply restarted.
Failures:
```
create_and_destroy: /obj/item/bodypart/arm/left/golem hard deleted 1 times out of a total del count of 7 at code/modules/unit_tests/create_and_destroy.dm:185
create_and_destroy: /obj/item/bodypart/arm/right/golem hard deleted 1 times out of a total del count of 7 at code/modules/unit_tests/create_and_destroy.dm:185
```

### PR DESCRIPTION
**Pull request by Jacquerel**
## About The Pull Request

Fixes #75251 
Fixes #75376
Fixes #75377 
Fixes #75394

uh... I think?

Here's the thing, is that golem arms are not actually particularly different to other kinds of arms, so I have been struggling to find a reason that this test is flaking _only_ for golem arms, in relation to a list that I didn't touch. 
My best guess is that the reason they're sometimes sticking to the "available hands" list is related to golems not being dismemberable and thereby skipping the code path which would remove a removed arm from the hand list, which sounds plausible except that nightmares are also NODISMEMBER and don't have the same problem.

Regardless, I made some boilerplate for when you give someone an arm which registers to the destruction and dismember signals and removes them from your available hand list when either of those things happen under any circumstances, whereas before it was reliant on passing through a much more specific pipeline.

Notably the text which set this also remarked `We only want to do this if the limb being removed is the active hand part.` which certainly does not sound true to me (although it does for the other line scoped inside that if check) and now also isn't, but this didn't seem to be widely affecting arms in general anyway.

I couldn't replicate the test failing after making these changes but as a flaky result that doesn't really tell us very much because to some extent it might just have been down to luck.

## Why It's Good For The Game

Memory management, hopefully will stop randomly triggering CI failures.

## Changelog

Not player facing

### CODE REVIEW

File: code/modules/mob/living/carbon/carbon.dm [hunk 1]
```diff
@@ -982,12 +982,24 @@
...
 				r_arm_index_next += 2
 				bodypart_instance.held_index = r_arm_index_next //2, 4, 6, 8...
-				hand_bodyparts += bodypart_instance
+				on_added_hand(bodypart_instance, r_arm_index_next)
 
+/// Called when a new hand is added
+/mob/living/carbon/proc/on_added_hand(obj/item/bodypart/arm/new_hand, hand_index)
+	if(hand_index > hand_bodyparts.len)
+		hand_bodyparts.len = hand_index
+	hand_bodyparts[hand_index] = new_hand
```
Comment (Mothblocks) [inline code/modules/mob/living/carbon/carbon.dm:995]: Do we loop over `hand_bodyparts` at any point? Will having nulls in there break things?

Comment (Jacquerel) [reply code/modules/mob/living/carbon/carbon.dm:995]: Having nulls is the expected behaviour. The index is frequently used directly as a lookup into this list (potentially a bad idea begging for runtimes) so removing things from it _without_ leaving a null would runtime, and also change the order of the list.

Comment (Mothblocks) [reply code/modules/mob/living/carbon/carbon.dm:995]: Okay, just making sure

**Review by Mothblocks — APPROVED** [review]
I do not like that you have no idea what the actual bug is, but this is also happening in like 30% of tests so

**Comment by LemonInTheDark** [comment]
> I do not like that you have no idea what the actual bug is, but this is also happening in like 30% of tests so

**Pull request merged**